### PR TITLE
core: fix NEO balance state handler

### PR DESCRIPTION
**Pull request by AnnaShaleva**
### Problem

States diff at 289026 block of testnet.

Transaction script:
```
NEO-GO-VM > loadbase64 CxEMFFNELbmZXXdH6APtO+RkGzpoWdjPDBQEuHjC2NkH7PHXGJkPkDhnxZKhExTAHwwIdHJhbnNmZXIMFPVj6kC8KD1NDgXEjqMFs/Kgc0DvQWJ9W1I5
READY: loaded 87 instructions
NEO-GO-VM 0 > ops
INDEX    OPCODE       PARAMETER                                   
0        PUSHNULL                                                 <<
1        PUSH1                                                    
2        PUSHDATA1    53442db9995d7747e803ed3be4641b3a6859d8cf    
24       PUSHDATA1    04b878c2d8d907ecf1d718990f903867c592a113    
46       PUSH4                                                    
47       PACK                                                     
48       PUSH15                                                   
49       PUSHDATA1    7472616e73666572 ("transfer")               
59       PUSHDATA1    f563ea40bc283d4d0e05c48ea305b3f2a07340ef    
81       SYSCALL      System.Contract.Call (627d5b52)             
86       ASSERT                                                   
```

C#'s applog:
```
{
   "jsonrpc" : "2.0",
   "result" : {
      "txid" : "0x04693267bd336469d0df9cb760a750808dc005ff5327df8b598f04d2f24b1c11",
      "executions" : [
         {
            "gasconsumed" : "16873950",
            "exception" : null,
            "trigger" : "Application",
            "stack" : [],
            "vmstate" : "HALT",
            "notifications" : [
               {
                  "contract" : "0xd2a4cff31913016155e38e474a2c06d08be276cf",
                  "state" : {
                     "value" : [
                        {
                           "type" : "Any"
                        },
                        {
                           "value" : "BLh4wtjZB+zx1xiZD5A4Z8WSoRM=",
                           "type" : "ByteString"
                        },
                        {
                           "type" : "Integer",
                           "value" : "68"
                        }
                     ],
                     "type" : "Array"
                  },
                  "eventname" : "Transfer"
               },
               {
                  "state" : {
                     "value" : [
                        {
                           "type" : "Any"
                        },
                        {
                           "value" : "U0QtuZldd0foA+075GQbOmhZ2M8=",
                           "type" : "ByteString"
                        },
                        {
                           "value" : "1",
                           "type" : "Integer"
                        }
                     ],
                     "type" : "Array"
                  },
                  "eventname" : "Transfer",
                  "contract" : "0xd2a4cff31913016155e38e474a2c06d08be276cf"
               },
               {
                  "contract" : "0xef4073a0f2b305a38ec4050e4d3d28bc40ea63f5",
                  "state" : {
                     "value" : [
                        {
                           "type" : "ByteString",
                           "value" : "U0QtuZldd0foA+075GQbOmhZ2M8="
                        },
                        {
                           "type" : "ByteString",
                           "value" : "ibZHTh0Sf77wu4Bk1QEh8JmH9pk="
                        },
                        {
                           "value" : "0",
                           "type" : "Integer"
                        }
                     ],
                     "type" : "Array"
                  },
                  "eventname" : "Transfer"
               },
               {
                  "eventname" : "Transfer",
                  "state" : {
                     "type" : "Array",
                     "value" : [
                        {
                           "value" : "BLh4wtjZB+zx1xiZD5A4Z8WSoRM=",
                           "type" : "ByteString"
                        },
                        {
                           "type" : "ByteString",
                           "value" : "U0QtuZldd0foA+075GQbOmhZ2M8="
                        },
                        {
                           "value" : "1",
                           "type" : "Integer"
                        }
                     ]
                  },
                  "contract" : "0xef4073a0f2b305a38ec4050e4d3d28bc40ea63f5"
               },
               {
                  "state" : {
                     "type" : "Array",
                     "value" : [
                        {
                           "type" : "ByteString",
                           "value" : "U0QtuZldd0foA+075GQbOmhZ2M8="
                        },
                        {
                           "value" : "ibZHTh0Sf77wu4Bk1QEh8JmH9pk=",
                           "type" : "ByteString"
                        },
                        {
                           "type" : "Integer",
                           "value" : "1"
                        }
                     ]
                  },
                  "eventname" : "Transfer",
                  "contract" : "0xef4073a0f2b305a38ec4050e4d3d28bc40ea63f5"
               }
            ]
         }
      ]
   },
   "id" : 1
}
```

Go's applog:
```
{
   "id" : 1,
   "jsonrpc" : "2.0",
   "result" : {
      "executions" : [
         {
            "gasconsumed" : "16874835",
            "exception" : "at instruction 92 (SYSCALL): failed native call: at instruction 92 (SYSCALL): failed native call: at instruction 92 (SYSCALL): failed native call: at instruction 4 (INITSLOT): gas limit is exceeded",
            "stack" : [
               {
                  "type" : "Any"
               },
               {
                  "value" : "1",
                  "type" : "Integer"
               },
               {
                  "type" : "Any"
               }
            ],
            "vmstate" : "FAULT",
            "trigger" : "Application",
            "notifications" : [
               {
                  "state" : {
                     "type" : "Array",
                     "value" : [
                        {
                           "type" : "Any"
                        },
                        {
                           "type" : "ByteString",
                           "value" : "BLh4wtjZB+zx1xiZD5A4Z8WSoRM="
                        },
                        {
                           "value" : "68",
                           "type" : "Integer"
                        }
                     ]
                  },
                  "contract" : "0xd2a4cff31913016155e38e474a2c06d08be276cf",
                  "eventname" : "Transfer"
               },
               {
                  "eventname" : "Transfer",
                  "contract" : "0xd2a4cff31913016155e38e474a2c06d08be276cf",
                  "state" : {
                     "type" : "Array",
                     "value" : [
                        {
                           "type" : "Any"
                        },
                        {
                           "value" : "U0QtuZldd0foA+075GQbOmhZ2M8=",
                           "type" : "ByteString"
                        },
                        {
                           "value" : "1",
                           "type" : "Integer"
                        }
                     ]
                  }
               },
               {
                  "contract" : "0xd2a4cff31913016155e38e474a2c06d08be276cf",
                  "state" : {
                     "type" : "Array",
                     "value" : [
                        {
                           "type" : "Any"
                        },
                        {
                           "value" : "U0QtuZldd0foA+075GQbOmhZ2M8=",
                           "type" : "ByteString"
                        },
                        {
                           "value" : "1",
                           "type" : "Integer"
                        }
                     ]
                  },
                  "eventname" : "Transfer"
               },
               {
                  "state" : {
                     "type" : "Array",
                     "value" : [
                        {
                           "type" : "Any"
                        },
                        {
                           "type" : "ByteString",
                           "value" : "U0QtuZldd0foA+075GQbOmhZ2M8="
                        },
                        {
                           "type" : "Integer",
                           "value" : "1"
                        }
                     ]
                  },
                  "contract" : "0xd2a4cff31913016155e38e474a2c06d08be276cf",
                  "eventname" : "Transfer"
               }
            ]
         }
      ],
      "txid" : "0x04693267bd336469d0df9cb760a750808dc005ff5327df8b598f04d2f24b1c11"
   }
}
```

### Solution

We need to store account's balance height update before GAS mint,
because mint can call OnNEP17Payment and OnNEP17Payment can call NEO
transfer which also calls GAS mint. Storing balance height allows to
avoid GAS minting recursion.

### CODE REVIEW

File: pkg/core/native/native_neo.go [hunk 1]
```diff
@@ -442,6 +442,15 @@ func (n *NEO) distributeGas(ic *interop.Context, h util.Uint160, acc *state.NEOB
 		return err
 	}
 	acc.BalanceHeight = ic.Block.Index
+
+	// Must store acc before GAS distribution to fix acc's BalanceHeight value in the storage for
+	// further acc's queries from `onNEP17Payment` if so.
+	key := makeAccountKey(h)
+	err = ic.DAO.PutStorageItem(n.ID, key, acc.Bytes())
```
Comment (roman-khimov) [inline pkg/core/native/native_neo.go:449]: What's interesting is that the balance value is actually going to be the old one.

Comment (AnnaShaleva) [reply pkg/core/native/native_neo.go:449]: One more interesting thing is that [LazyCoin contract](https://github.com/vang1ong7ang/Tanya/blob/main/PoC/InsufficientFunds.cs#L29) should own some NEO for a time period to trigger GAS mint inside `onNEP17Payment`. But the whole NEO balance is transferred to the owner each time onNEP17Payment is called.

File: pkg/core/helper_test.go [hunk 2]
```diff
@@ -395,6 +396,7 @@ func initBasicChain(t *testing.T, bc *Blockchain) {
 		},
 	}
 	require.NoError(t, addNetworkFee(bc, transferTx, acc0))
+	transferTx.SystemFee += 1000000
```
Comment (roman-khimov) [inline pkg/core/helper_test.go:399]: Why? We already have `11000000` system fee and the change here shouldn't affect the amount consumed for simple transfer.

File: pkg/core/helper_test.go [hunk 3]
```diff
@@ -603,14 +606,25 @@ func prepareContractMethodInvokeGeneric(chain *Blockchain, sysfee int64,
 func signTxWithAccounts(chain *Blockchain, tx *transaction.Transaction, accs ...*wallet.Account) {
 	scope := transaction.CalledByEntry
 	for _, acc := range accs {
+		var accH util.Uint160
+		if acc.Contract.Deployed {
+			accH, _ = address.StringToUint160(acc.Address)
```
Comment (roman-khimov) [inline pkg/core/helper_test.go:611]: Can we always get it from `acc.Address`?

File: pkg/core/native_neo_test.go [hunk 4]
```diff
@@ -195,6 +196,65 @@ func TestNEO_Vote(t *testing.T) {
...
+	bc := newTestChain(t)
+	initBasicChain(t, bc)
+
+	contractHash, err := bc.GetContractScriptHash(1) // deployed rpc/server/testdata/test_contract.go contract
+	require.NoError(t, err)
+	tx := transferTokenFromMultisigAccount(t, bc, contractHash, bc.contracts.GAS.Hash, 2_0000_0000)
+	checkTxHalt(t, bc, tx.Hash())
+
+	// Transfer 1 NEO to test contract, the contract should earn some GAS by owning this NEO.
+	tx = transferTokenFromMultisigAccount(t, bc, contractHash, bc.contracts.NEO.Hash, 1)
```
Comment (roman-khimov) [inline pkg/core/native_neo_test.go:210]: If you transfer 10 here...

File: pkg/core/native_neo_test.go [hunk 5]
```diff
@@ -195,6 +196,65 @@ func TestNEO_Vote(t *testing.T) {
...
+	require.Equal(t, vm.HaltState, res[0].VMState)
+
+	acc := &wallet.Account{
+		Address: address.Uint160ToString(contractHash),
+		Contract: &wallet.Contract{
+			Deployed: true,
+		},
+	}
+
+	// Register candidate and vote for this candidate by contract to earn some GAS.
```
Comment (roman-khimov) [inline pkg/core/native_neo_test.go:222]: ...you might as well avoid voting here (which simplifies test).

**Comment by roman-khimov** [comment]
I forgot to mention, I've quickly tested it for performance and there is no noticeable changes compared to master.

File: pkg/core/helper_test.go (hunk 2 above)
Comment (AnnaShaleva) [reply pkg/core/helper_test.go:399]: Let me see, not fixed yet, but the previously set amount isn't enough.

Comment (AnnaShaleva) [reply pkg/core/helper_test.go:399]: It's connected with onNEP17Payment method added to a contract (which is the sender).

**Review by roman-khimov — APPROVED** [review]

File: pkg/core/helper_test.go [hunk 6]
```diff
@@ -603,14 +606,20 @@ func prepareContractMethodInvokeGeneric(chain *Blockchain, sysfee int64,
 func signTxWithAccounts(chain *Blockchain, tx *transaction.Transaction, accs ...*wallet.Account) {
 	scope := transaction.CalledByEntry
 	for _, acc := range accs {
+		accH, _ := address.StringToUint160(acc.Address)
 		tx.Signers = append(tx.Signers, transaction.Signer{
-			Account: acc.PrivateKey().GetScriptHash(),
+			Account: accH,
```
Comment (fyrchik) [inline pkg/core/helper_test.go:611]: `acc.Contract.ScriptHash()` should also work for both cases.

File: pkg/core/helper_test.go (hunk 2 above)
Comment (AnnaShaleva) [reply pkg/core/helper_test.go:399]: Got this, I've added the contract's owner:
```
var owner = util.FromAddress("NbrUYaZgyhSkNoRo9ugRyEMdUZxrhkNaWB")
```
so it needs to be initialised before transfer invocation. Previosly set amount is not enough for initialisation.

File: pkg/core/helper_test.go (hunk 6 above)
Comment (AnnaShaleva) [reply pkg/core/helper_test.go:611]: Isn't `acc.Contract.ScriptHash()` a hash of contract's script? Here I need the contract address (hash) in the chain, and now it's not the same as the hash of the contract script.

Comment (fyrchik) [reply pkg/core/helper_test.go:611]: You are right.

**Pull request merged**